File: src/futurnal/cli/activity.py
"""Activity Stream CLI commands.

Step 08: Frontend Intelligence Integration - Phase 3

Provides unified activity stream combining:
- Audit logs (system events)
- Search history
- Chat sessions
- Ingestion events
- Schema evolution

Research Foundation:
- AgentFlow: Activity tracking patterns
- RLHI: User interaction history
"""

import json
import logging
from datetime import datetime, timedelta
from pathlib import Path
from typing import List, Optional, Dict, Any

from typer import Typer, Option

logger = logging.getLogger(__name__)

activity_app = Typer(help="Activity stream commands")


def _get_workspace_dir() -> Path:
    """Get Futurnal workspace directory."""
    import os
    return Path(os.path.expanduser("~/.futurnal"))


def _read_audit_logs(
    workspace: Path,
    limit: int,
    date_from: Optional[datetime],
    date_to: Optional[datetime],
) -> List[Dict[str, Any]]:
    """Read events from audit logs."""
    events = []
    audit_path = workspace / "audit" / "audit.log"

    if not audit_path.exists():
        return events

    try:
        with open(audit_path, 'r') as f:
            lines = f.readlines()

        # Process most recent first
        for line in reversed(lines[-1000:]):  # Cap at 1000 lines for performance
            if len(events) >= limit:
                break

            try:
                entry = json.loads(line.strip())
                timestamp_str = entry.get("timestamp")
                if not timestamp_str:
                    continue

                timestamp = datetime.fromisoformat(timestamp_str.replace("Z", "+00:00"))

                # Apply date filters
                if date_from and timestamp < date_from:
                    continue
                if date_to and timestamp > date_to:
                    continue

                # Map to activity event
                event_type = _map_action_to_type(entry.get("action", ""))
                event = {
                    "id": entry.get("job_id", f"audit-{timestamp.timestamp()}"),
                    "type": event_type,
                    "category": _get_event_category(event_type),
                    "title": _format_audit_title(entry),
                    "description": entry.get("redacted_path") or entry.get("metadata", {}).get("message"),
                    "timestamp": timestamp.isoformat(),
                    "relatedEntityIds": [],
                    "metadata": {
                        "source": entry.get("source"),
                        "action": entry.get("action"),
                        "status": entry.get("status"),
                    },
                }
                events.append(event)

            except (json.JSONDecodeError, ValueError) as e:
                continue

    except Exception as e:
        logger.warning(f"Failed to read audit logs: {e}")

    return events


def _map_action_to_type(action: str) -> str:
    """Map audit action to activity type."""
    action_lower = action.lower()

    if "search" in action_lower:
        return "search"
    elif "chat" in action_lower:
        return "chat"
    elif "ingest" in action_lower or "process" in action_lower:
        return "document"
    elif "consent" in action_lower:
        return "schema"
    elif "entity" in action_lower or "graph" in action_lower:
        return "entity"
    elif "learn" in action_lower or "pattern" in action_lower:
        return "insight"
    else:
        return "document"


def _get_event_category(event_type: str) -> str:
    """Get category for event type."""
    if event_type in ("search", "chat"):
        return "interaction"
    elif event_type in ("insight", "schema"):
        return "learning"
    else:
        return "data"


def _format_audit_title(entry: Dict[str, Any]) -> str:
    """Format audit entry as activity title."""
    action = entry.get("action", "Unknown action")
    source = entry.get("source", "system")
    status = entry.get("status", "")

    # Clean up action for display
    action = action.replace("_", " ").replace(":", " - ").title()

    if status:
        return f"{action} ({status})"
    return action


@activity_app.command("list")
def list_activities(
    limit: int = Option(50, "--limit", "-l", help="Maximum number of activities to return"),
    offset: int = Option(0, "--offset", "-o", help="Offset for pagination"),
    event_types: Optional[str] = Option(None, "--types", "-t", help="Comma-separated event types filter"),
    date_from: Optional[str] = Option(None, "--from", "-f", help="Start date (ISO format)"),
    date_to: Optional[str] = Option(None, "--to", help="End date (ISO format)"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """List activity events from all sources.

    Combines audit logs, search history, chat sessions, and more.

    Examples:
        futurnal activity list
        futurnal activity list --limit 20 --types search,chat
        futurnal activity list --from 2024-01-01 --json
    """
    try:
        workspace = _get_workspace_dir()

        # Parse date filters
        parsed_from = datetime.fromisoformat(date_from) if date_from else None
        parsed_to = datetime.fromisoformat(date_to) if date_to else None

        # Parse event types filter
        type_filter = event_types.split(",") if event_types else None

        # Collect events from all sources
        all_events = []

        # 1. Audit logs
        audit_events = _read_audit_logs(
            workspace,
            limit=limit * 2,  # Fetch more, filter later
            date_from=parsed_from,
            date_to=parsed_to,
        )
        all_events.extend(audit_events)

        # 2. Additional sources could be added here
        # - Search history from searchStore
        # - Chat sessions from chatStore
        # - etc.

        # Sort by timestamp (newest first)
        all_events.sort(key=lambda e: e.get("timestamp", ""), reverse=True)

        # Apply type filter
        if type_filter:
            all_events = [e for e in all_events if e.get("type") in type_filter]

        # Apply pagination
        paginated = all_events[offset:offset + limit]

        if output_json:
            output = {
                "success": True,
                "events": paginated,
                "total": len(all_events),
                "limit": limit,
                "offset": offset,
            }
            print(json.dumps(output))
        else:
            print(f"Activity Stream ({len(paginated)} of {len(all_events)} events)")
            print("-" * 50)

            for event in paginated:
                ts = event.get("timestamp", "")
                try:
                    ts_display = datetime.fromisoformat(ts.replace("Z", "+00:00")).strftime("%Y-%m-%d %H:%M")
                except:
                    ts_display = ts[:16]

                type_icon = {
                    "search": "[S]",
                    "chat": "[C]",
                    "document": "[D]",
                    "entity": "[E]",
                    "insight": "[I]",
                    "schema": "[+]",
                }.get(event.get("type", ""), "[?]")

                print(f"{type_icon} {ts_display} | {event.get('title', 'Unknown')}")

    except Exception as e:
        logger.exception("Failed to list activities")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")
        raise SystemExit(1)


@activity_app.command("recent")
def recent_activities(
    limit: int = Option(10, "--limit", "-l", help="Number of recent activities"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Get most recent activities (shortcut for list --limit N).

    Examples:
        futurnal activity recent
        futurnal activity recent --limit 5 --json
    """
    # Just call list with appropriate defaults
    list_activities(
        limit=limit,
        offset=0,
        event_types=None,
        date_from=None,
        date_to=None,
        output_json=output_json,
    )

File: desktop/src-tauri/src/commands/activity.rs
//! Activity Stream IPC commands.
//!
//! Step 08: Frontend Intelligence Integration - Phase 3
//!
//! Research Foundation:
//! - AgentFlow: Activity tracking patterns
//! - RLHI: User interaction history
//!
//! Provides activity stream from all sources:
//! - Audit logs
//! - Search history
//! - Chat sessions
//! - Ingestion events

use serde::{Deserialize, Serialize};
use tauri::command;

/// Activity event from the stream.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityEvent {
    pub id: String,
    #[serde(rename = "type")]
    pub event_type: String,
    pub category: String,
    pub title: String,
    pub description: Option<String>,
    pub timestamp: String,
    pub related_entity_ids: Vec<String>,
    pub metadata: serde_json::Value,
}

/// Response from activity list command.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityListResponse {
    pub success: bool,
    pub events: Vec<ActivityEvent>,
    pub total: i32,
    pub limit: i32,
    pub offset: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Get activity stream with optional filters.
///
/// Step 08: Activity Stream
///
/// Calls: `futurnal activity list --json`
#[command]
pub async fn get_activity_log(
    limit: Option<i32>,
    offset: Option<i32>,
    event_types: Option<Vec<String>>,
    date_from: Option<String>,
    date_to: Option<String>,
) -> Result<ActivityListResponse, String> {
    let limit_str = limit.unwrap_or(50).to_string();
    let offset_str = offset.unwrap_or(0).to_string();

    let mut args = vec![
        "activity".to_string(),
        "list".to_string(),
        "--limit".to_string(),
        limit_str,
        "--offset".to_string(),
        offset_str,
        "--json".to_string(),
    ];

    // Add event types filter
    if let Some(types) = event_types {
        if !types.is_empty() {
            args.push("--types".to_string());
            args.push(types.join(","));
        }
    }

    // Add date filters
    if let Some(from) = date_from {
        args.push("--from".to_string());
        args.push(from);
    }
    if let Some(to) = date_to {
        args.push("--to".to_string());
        args.push(to);
    }

    let args_refs: Vec<&str> = args.iter().map(|s| s.as_str()).collect();

    let response: ActivityListResponse = match crate::python::execute_cli(&args_refs).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Activity list CLI failed: {}", e);
            return Ok(ActivityListResponse {
                success: false,
                events: vec![],
                total: 0,
                limit: limit.unwrap_or(50),
                offset: offset.unwrap_or(0),
                error: Some(format!("Activity list failed: {}", e)),
            });
        }
    };

    log::info!(
        "Retrieved {} activities (total: {})",
        response.events.len(),
        response.total
    );

    Ok(response)
}

/// Get recent activities (shortcut).
///
/// Calls: `futurnal activity recent --json`
#[command]
pub async fn get_recent_activities(
    limit: Option<i32>,
) -> Result<ActivityListResponse, String> {
    get_activity_log(limit, Some(0), None, None, None).await
}
